packages/ember-data/lib/system/model/model.js

```diff
@@ -781,7 +782,11 @@ var Model = Ember.Object.extend(Ember.Evented, {
 
   hasMany: function() {
     Ember.assert("The `hasMany` method is not available on DS.Model, a DS.Snapshot was probably expected. Are you passing a DS.Model instead of a DS.Snapshot to your serializer?", false);
-  }
+  },
+
+  setId: Ember.observer('id', function () {
```

Comment:
if I'm understanding you correctly, that was the first attempt with the whole object.defineproperty thing. That didn't work because of a bug in ember itself.